engine/src/main/java/org/terasology/rendering/world/WorldRendererLwjgl.java

```diff
@@ -249,23 +236,36 @@ private void preRenderUpdate(DefaultRenderingProcess.StereoRenderState stereoRen
             renderableWorld.update();
             renderableWorld.generateVBOs();
             secondsSinceLastFrame = 0;
+
+            updateMaterials(); // TODO: this should happen only if necessary, i.e. upon notification from the Asset system.
         }
 
-        if (stereoRenderState != DefaultRenderingProcess.StereoRenderState.MONO) {
+        if (currentRenderingStage != WorldRenderingStage.MONO) {
             playerCamera.updateFrustum();
         }
 
-        // this line needs to be here as deep down it relies on the camera's frustrum.
+        // this line needs to be here as deep down it relies on the camera's frustrum, updated just above.
         renderableWorld.queueVisibleChunks(isFirstRenderingStageForCurrentFrame);
     }
 
     /**
+     * This allows for materials to be swapped at runtime, i.e. if a module provides new ones.
+     */
+    private void updateMaterials() {
+        // TODO: review - perhaps only the chunk shader would ever need to be swapped at runtime? All the others could be final?
+        chunkShader         = Assets.getMaterial("engine:prog.chunk");
+        lightGeometryShader = Assets.getMaterial("engine:prog.lightGeometryPass");
+        //simpleShader        = Assets.getMaterial("engine:prog.simple");  // in use by the currently commented out light stencil pass
+        shadowMapShader     = Assets.getMaterial("engine:prog.shadowMap");
+    }
```

Comment:
Well, previously references to some of these shaders were allocated and obtained with Assets.getMaterial() every time the renderChunk() method run. 

I interpreted this to support the eventuality that the shaders might change at runtime, i.e. because a module overrides them. So I moved the reference fetching out of renderChunk(), but the materials still get fetched from Assets once per frame. Ideally they'd be fetched again only when necessary. 

From what you write though, I can at least make an isDisposed() check to avoid the getMaterial() call in most cases. Or should I just get the materials in the renderer's constructor to never touch them again?